### PR TITLE
Fix ReferenceTerminals and SlackTerminal iIDM extension with merged networks and subnetworks

### PR DESCRIPTION
**Pull request by jeandemanged**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Bug fix

**What is the current behavior?**

ReferenceTerminals and SlackTerminal iIDM extensions were never revisited to work properly with subnetworks.
In particular:
- We get an exception when merging a network having a ReferenceTerminals extension, writing it, and reading it.
- `ReferenceTerminals.reset` and `SlackTerminal.reset` do not reset the subnetworks

**What is the new behavior (if this is a feature change)?**
ReferenceTerminals fixed to properly check the network owning the terminal.
`reset` methods will also reset the subnetworks.

**Does this PR introduce a breaking change or deprecate an API?**
- [x] No

### CODE REVIEW

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java [hunk 1]
```diff
@@ -39,7 +40,7 @@ default String getName() {
     ReferenceTerminals addReferenceTerminal(Terminal terminal);
 
     /**
-     * Deletes all defined reference terminals in the network for the current variant
+     * Deletes all defined reference terminals in the network and all its subnetworks for the current variant
```
Comment (jeandemanged) [inline iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:43]: this is what e.g. OLF will want do

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java [hunk 2]
```diff
@@ -51,10 +52,13 @@ static void reset(Network network) {
...
         }
         ext.reset();
+        // reset also all subnetwork
+        network.getSubnetworks().forEach(ReferenceTerminals::reset);
     }
 
     /**
-     * Defines/add a terminal as reference in the network for the current variant
+     * Defines/add a terminal as reference in the network for the current variant.
+     * In case of a merged network with subnetwork, the extension is placed on the root/merged network.
```
Comment (jeandemanged) [inline iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:61]: this is what e.g. OLF will want do

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java [hunk 3]
```diff
@@ -70,15 +74,25 @@ static void addTerminal(Terminal terminal) {
     }
 
     /**
-     * Gets the reference terminals in the network for the current variant
+     * Gets the reference terminals in the network for the current variant.
+     * In case of a merged network with subnetworks, reference terminals from the merged network
+     * and reference terminals from subnetworks are returned altogether.
```
Comment (jeandemanged) [inline iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:79]: this is what eg CGMES SV exporter will want to do

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java (hunk 2 above)
Comment (annetill) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:61]: Ok for me.

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java (hunk 1 above)
Comment (annetill) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:43]: It depends on which network the load flow is run. If you run it on a merged network, you can/have to leave the reference terminals of the subnetworks.

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java (hunk 3 above)
Comment (annetill) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:79]: Not sure about this. IGM SV export: the extension from the subnetwork, CGM SV export, the extension from the merged network only.

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java (hunk 1 above)
Comment (annetill) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:43]: static void reset(Network network) {
    network.getVoltageLevels().forEach((vl) -> {
        reset(vl, (Terminal)null);
    });
}

static void reset(VoltageLevel voltageLevel, Terminal terminal) {
    Objects.requireNonNull(voltageLevel);
    SlackTerminal st = (SlackTerminal)voltageLevel.getExtension(SlackTerminal.class);
    if (st == null && terminal != null) {
        ((SlackTerminalAdder)voltageLevel.newExtension(SlackTerminalAdder.class)).withTerminal(terminal).add();
    } else if (st != null) {
        st.setTerminal(terminal, true);
    }

}  
-> we need also to align here if we clean the subnetworks too.

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java (hunk 3 above)
Comment (jeandemanged) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java]: I reverted this change

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java (hunk 1 above)
Comment (jeandemanged) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java:42]: done / SlackTerminal.reset modified to reset also the subnetworks.
also updated PR title & description

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/ReferenceTerminalsImpl.java [hunk 4]
```diff
@@ -39,12 +39,14 @@ public ReferenceTerminalsImpl(Network network, Set<Terminal> terminals) {
                 Collections.nCopies(getVariantManagerHolder().getVariantManager().getVariantArraySize(), new LinkedHashSet<>()));
         setReferenceTerminals(terminals);
         this.referenceTerminalsListener = new ReferenceTerminalsListener();
-        network.addListener(this.referenceTerminalsListener);
+        // listeners are not (yet) supported on subnetworks
+        network.getNetwork().addListener(this.referenceTerminalsListener);
```
Comment (olperr1) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/ReferenceTerminalsImpl.java:43]: I think that there is a problem if the extension is in a network **before** it is merged in another one.
In `AbstractNetwork.transferExtensions(…)`, the extension will be transferred from the `NetworkImpl` object (the network to merge) to the `SubnetworkImpl` (the merged subnetwork) via `Extension.setExtendable(…)`.
The listener should also be transferred from the old extendable to the new one in this method.

**Review by olperr1 — CHANGES_REQUESTED** [review]

File: iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/extensions/AbstractReferenceTerminalsTest.java [hunk 5]
```diff
@@ -196,6 +196,55 @@ public void testWrongNetwork() {
         assertEquals("Terminal given is not in the right Network (sim1 instead of fourSubstations)", ex2.getMessage());
     }
 
+    @Test
+    public void testWithSubnetwork() {
+        Network merged = Network.merge(network, EurostagTutorialExample1Factory.create());
```
Comment (olperr1) [inline iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/extensions/AbstractReferenceTerminalsTest.java:201]: Could you add unit tests using the listener?

**Review by olperr1 — CHANGES_REQUESTED** [review]

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java [hunk 6]
```diff
@@ -70,7 +73,10 @@ static void addTerminal(Terminal terminal) {
     }
 
     /**
-     * Gets the reference terminals in the network for the current variant
+     * Gets the reference terminals in the network for the current variant.
+     * <p> Note: This method returns only the terminal from the extension attached to the provided network.
+     * In case of a merged network with subnetworks, be careful whether you want the extension
+     * of the merged network or of a subnetwork.
```
[the remark this answers is not in the document]
Comment (jeandemanged) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/ReferenceTerminals.java]: thanks, done

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/ReferenceTerminalsImpl.java (hunk 4 above)
Comment (jeandemanged) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/ReferenceTerminalsImpl.java]: Yes indeed there was a problem I was aware of, and I thought if was just impossible to solve, before you showed me the way, thanks again!

File: iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/extensions/AbstractReferenceTerminalsTest.java (hunk 5 above)
Comment (jeandemanged) [reply iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/extensions/AbstractReferenceTerminalsTest.java:201]: added a specific test with merge and detach

**Review by flo-dup — APPROVED** [review]

File: iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/extensions/AbstractReferenceTerminalsTest.java (hunk 5 above)
Comment (olperr1) [reply iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/extensions/AbstractReferenceTerminalsTest.java:201]: Thanks!

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/ReferenceTerminalsImpl.java (hunk 4 above)
Comment (olperr1) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/ReferenceTerminalsImpl.java]: Sorry, I hadn't understood that you encountered a problem here.
I'm glad the way we discussed together worked.
Thanks for the changes.

**Review by olperr1 — APPROVED** [review]

**Pull request merged**